### PR TITLE
[AIRFLOW-3231] Basic operators for Google Cloud SQL

### PR DESCRIPTION
**Pull request by sprzedwojski**
### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW-3231) issue and references it in the PR title.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

Add CloudSqlInstanceInsertOperator, CloudSqlInstancePatchOperator and CloudSqlInstanceDeleteOperator.

Each operator includes:
- core logic
- input params validation
- unit tests
- presence in the example DAG
- docstrings
- How-to and Integration documentation

Additionally, small improvements to GcpBodyFieldValidator were made:
- add simple list validation capability (type="list")
- introduced parameter allow_empty, which can be set to False
	to test for non-emptiness of a string instead of specifying
	a regexp.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

`test_gcp_sql_operator.py`

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality

- [x] Passes `flake8`

Co-authored-by: sprzedwojski <szymon.przedwojski@polidea.com>
Co-authored-by: potiuk <jarek.potiuk@polidea.com>

### CODE REVIEW

File: airflow/contrib/operators/gcp_sql_operator.py [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+from airflow.contrib.hooks.gcp_sql_hook import CloudSqlHook
+from airflow.contrib.utils.gcp_field_validator import GcpBodyFieldValidator
+from airflow.models import BaseOperator
+from airflow.utils.decorators import apply_defaults
+
+SETTINGS = 'settings'
+SETTINGS_VERSION = 'settingsVersion'
+
+CLOUD_SQL_VALIDATION = [
+    dict(name="name", allow_empty=False),
```
Comment (kaxil) [inline airflow/contrib/operators/gcp_sql_operator.py:31]: This really looks clunky... is there a way we can simplify this?

File: airflow/contrib/operators/gcp_sql_operator.py [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+            dict(name="verifyServerCertificate", optional=True)
+        ], optional=True),
+    ], optional=True)
+]
+
+
+class CloudSqlBaseOperator(BaseOperator):
+    """
+    Abstract base operator for Google Cloud SQL operators to inherit from.
+    """
```
Comment (kaxil) [inline airflow/contrib/operators/gcp_sql_operator.py:108]: Can we please add docstrings for this operator

Comment (kaxil) [reply airflow/contrib/operators/gcp_sql_operator.py:108]: I know it is a base operator but still if someone wants to implement something custom inherting this operator, it comes in handy to have some docstrings

File: airflow/contrib/operators/gcp_sql_operator.py [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+    :param gcp_conn_id: The connection ID used to connect to Google Cloud Platform.
+    :type gcp_conn_id: str
+    :param api_version: API version used (e.g. v1).
+    :type api_version: str
+    :param validate_body: True if body should be validated, False otherwise.
+    :type validate_body: bool
+    """
+    # [START gcp_sql_insert_template_fields]
+    template_fields = ('project_id', 'instance', 'gcp_conn_id', 'api_version')
+    # [END gcp_sql_insert_template_fields]
```
Comment (kaxil) [inline airflow/contrib/operators/gcp_sql_operator.py]: I like this

File: docs/howto/operator.rst [hunk 4]
```diff
@@ -283,3 +283,155 @@ See `Adding the IAM service agent user role to the runtime service <https://clou
...
+""""""""""""""""""
+
+.. literalinclude:: ../../airflow/contrib/example_dags/example_gcp_sql.py
+    :language: python
+    :dedent: 4
+    :start-after: [START howto_operator_cloudsql_delete]
+    :end-before: [END howto_operator_cloudsql_delete]
+
+Templating
+""""""""""
```
Comment (kaxil) [inline docs/howto/operator.rst:315]: Nice

**Review by kaxil — APPROVED** [review]
LGTM. Left 1-2 minor comments

File: airflow/contrib/hooks/gcp_sql_hook.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        :type instance: str
+        :return: A Cloud SQL instance resource.
+        :rtype: dict
+        """
+        return self.get_conn().instances().get(
+            project=project_id,
+            instance=instance
+        ).execute(num_retries=NUM_RETRIES)
+
+    def insert_instance(self, project_id, body):
```
Comment (kaxil) [inline airflow/contrib/hooks/gcp_sql_hook.py]: I know the api call's it `insert` but can we not call it `create` which is more easy to understand as we have in bigquery?

cc @fenglu-g

Comment (fenglu-db) [reply airflow/contrib/hooks/gcp_sql_hook.py]: "create" sounds good to me.

File: airflow/contrib/operators/gcp_sql_operator.py (hunk 2 above)
Comment (sprzedwojski) [reply airflow/contrib/operators/gcp_sql_operator.py:108]: Done

File: airflow/contrib/operators/gcp_sql_operator.py (hunk 3 above)
Comment (sprzedwojski) [reply airflow/contrib/operators/gcp_sql_operator.py]: Thanks, it was Jarek's idea ;)

File: airflow/contrib/hooks/gcp_sql_hook.py (hunk 5 above)
Comment (sprzedwojski) [reply airflow/contrib/hooks/gcp_sql_hook.py]: Done, I've renamed the operator to `CloudSqlInstanceCreateOperator` and also changed `insert` to `create` in all methods, comments, documentation etc.

File: airflow/contrib/operators/gcp_sql_operator.py (hunk 1 above)
Comment (sprzedwojski) [reply airflow/contrib/operators/gcp_sql_operator.py:31]: I discussed this a lot with others and it seems really hard to simplify at the moment while preserving the full validation functionality.

How about we leave it as is for now, given that such specs (although shorter) have already been merged in the past (for GCF and GCE operators), and we discuss the general approach to validation at a later time on the mailing list / Slack?

If we reach a conclusion I'll be happy to revisit all the merged specs and refactor them.

**Review by kaxil — APPROVED** [review]

**Pull request merged**